File: graph-algorithms/traversal/js/src/traversal.test.js
import {
	breadthFirstTraversal,
	depthFirstTraversal,
	depthFirstTraversalRec,
} from './traversal';

describe('depthFirstTraversal', () => {
	describe('given a graph', () => {
		/*
		 * a------→c
		 * |       |
		 * ↓       ↓
		 * b       e
		 * |
		 * ↓
		 * d------→f
		 */

		it('should traverse the graph depth first', () => {
			// Arrange.
			const graph = {
				a: ['c', 'b'],
				b: ['d'],
				c: ['e'],
				d: ['f'],
				e: [],
				f: [],
			}
			const expected = 'abdfce'

			// Act.
			const actual = depthFirstTraversal(graph, 'a');

			// Assert.
			expect(actual.join('')).toEqual(expected);
		});
	});
})

describe('depthFirstTraversalRec', () => {
	describe('given a graph', () => {
		/*
		 * a------→c
		 * |       |
		 * ↓       ↓
		 * b       e
		 * |
		 * ↓
		 * d------→f
		 */

		it('should traverse the graph depth first recursively', () => {
			// Arrange.
			const graph = {
				a: ['b', 'c'],
				b: ['d'],
				c: ['e'],
				d: ['f'],
				e: [],
				f: [],
			}
			const expected = 'abdfce'

			// Act.
			const actual = depthFirstTraversalRec(graph, 'a');

			// Assert.
			expect(actual.join('')).toEqual(expected);
		});
	});
})

describe('breadthFirstTraversal', () => {
	describe('given a graph', () => {
		/*
		 * a------→c
		 * |       |
		 * ↓       ↓
		 * b       e
		 * |
		 * ↓
		 * d------→f
		 */

		it('should traverse the graph breadth first', () => {
			// Arrange.
			const graph = {
				a: ['b', 'c'],
				b: ['d'],
				c: ['e'],
				d: ['f'],
				e: [],
				f: [],
			}
			const expected = 'abcdef'

			// Act.
			const actual = breadthFirstTraversal(graph, 'a');

			// Assert.
			expect(actual.join('')).toEqual(expected);
		});
	});
})

File: graph-algorithms/traversal/go/pkg/traversal/traversal_test.go
package traversal_test

import (
	"fmt"
	"strings"
	"testing"

	"github.com/stretchr/testify/assert"

	"github.com/claudemuller/algorithms/graph-algorithms/go/pkg/traversal"
)

func TestDepthFirstTraversal(t *testing.T) {
	type test struct {
		name  string
		src   string
		graph map[string][]string
		want  string
	}
	/*
	 * a------→c
	 * |       |
	 * ↓       ↓
	 * b       e
	 * |
	 * ↓
	 * d------→f
	 */
	tests := []test{
		{
			name: "depth first success",
			src:  "a",
			graph: map[string][]string{
				"a": {"c", "b"},
				"b": {"d"},
				"c": {"e"},
				"d": {"f"},
				"e": {},
				"f": {},
			},
			want: "abdfce",
		},
	}

	for _, tc := range tests {
		got := traversal.DepthFirstTraversal(tc.graph, tc.src)

		assert.Equal(t, tc.want, strings.Join(got, ""), fmt.Sprintf("%s: %s", tc.name, "depth first traversal failed"))
	}
}

func TestDepthFirstTraversalRec(t *testing.T) {
	type test struct {
		name  string
		src   string
		graph map[string][]string
		want  string
	}
	/*
	 * a------→c
	 * |       |
	 * ↓       ↓
	 * b       e
	 * |
	 * ↓
	 * d------→f
	 */
	tests := []test{
		{
			name: "depth first rec success",
			src:  "a",
			graph: map[string][]string{
				"a": {"b", "c"},
				"b": {"d"},
				"c": {"e"},
				"d": {"f"},
				"e": {},
				"f": {},
			},
			want: "abdfce",
		},
	}

	for _, tc := range tests {
		got := traversal.DepthFirstTraversalRec(tc.graph, tc.src)

		assert.Equal(t, tc.want, strings.Join(got, ""), fmt.Sprintf("%s: %s", tc.name, "depth first traversal failed"))
	}
}

func TestBreadthFirstTraversal(t *testing.T) {
	type test struct {
		name  string
		src   string
		graph map[string][]string
		want  string
	}
	/*
	 * a------→c
	 * |       |
	 * ↓       ↓
	 * b       e
	 * |
	 * ↓
	 * d------→f
	 */
	tests := []test{
		{
			name: "breadth first success",
			src:  "a",
			graph: map[string][]string{
				"a": {"b", "c"},
				"b": {"d"},
				"c": {"e"},
				"d": {"f"},
				"e": {},
				"f": {},
			},
			want: "abcdef",
		},
	}

	for _, tc := range tests {
		got := traversal.BreadthFirstTraversal(tc.graph, tc.src)

		assert.Equal(t, tc.want, strings.Join(got, ""), fmt.Sprintf("%s: %s", tc.name, "breadth first traversal failed"))
	}
}

func TestEdgeListToAdjList(t *testing.T) {
	type test struct {
		name string
		src  string
		list [][]string
		want map[string][]string
	}
	/*
	 * a------→c
	 * |       |
	 * ↓       ↓
	 * b       e
	 * |
	 * ↓
	 * d------→f
	 */
	tests := []test{
		{
			name: "edge list to adj list success",
			src:  "a",
			list: [][]string{
				{"a", "c"},
				{"a", "b"},
				{"c", "e"},
				{"b", "d"},
				{"d", "f"},
			},
			want: map[string][]string{
				"a": {"c", "b"},
				"b": {"a", "d"},
				"c": {"a", "e"},
				"d": {"b", "f"},
				"e": {"c"},
				"f": {"d"},
			},
		},
	}

	for _, tc := range tests {
		got := traversal.EdgeListToAdjList(tc.list)

		assert.Equal(t, tc.want, got, fmt.Sprintf("%s: %s", tc.name, "list conversion failed"))
	}
}
